File: solutions/ruby/phone-number/1/phone_number.rb
class PhoneNumber
  def self.clean(number)
    # 1. Extraemos solo los dígitos numéricos
    digits = number.gsub(/\D/, '')

    # 2. Manejamos el código de país (11 dígitos empezando por 1)
    if digits.length == 11 && digits.start_with?('1')
      digits = digits[1..-1]
    end

    # 3. Validamos la longitud final de 10 dígitos
    return nil unless digits.length == 10

    # 4. Validamos las reglas NXX (Código de área e intercambio entre 2-9)
    # El primer dígito (área) y el cuarto dígito (intercambio) no pueden ser 0 o 1
    return nil if digits[0].to_i < 2 || digits[3].to_i < 2

    # 5. Si pasa todas las pruebas, devolvemos el número limpio
    digits
  end
end

File: solutions/typescript/phone-number/1/phone-number.ts
export function clean(phoneNumber: string): string {
  // 1. Validar letras y puntuación ANTES de limpiar
  // Los tests esperan estos mensajes específicos si encuentran caracteres prohibidos
  if (/[a-zA-Z]/.test(phoneNumber)) {
    throw new Error('Letters not permitted');
  }
  if (/[@:!]/.test(phoneNumber)) {
    throw new Error('Punctuations not permitted');
  }

  // 2. Extraer solo los dígitos
  let digits = phoneNumber.replace(/\D/g, '');

  // 3. Validaciones de longitud y código de país
  if (digits.length < 10) {
    throw new Error('Must not be fewer than 10 digits');
  }

  if (digits.length === 11) {
    if (digits[0] !== '1') {
      throw new Error('11 digits must start with 1');
    }
    digits = digits.slice(1);
  }

  if (digits.length > 11) {
    throw new Error('Must not be greater than 11 digits');
  }
  
  if (digits.length !== 10) {
    // Caso para cuando después de procesar el 11 inicial, algo salió mal
    throw new Error('Incorrect number of digits');
  }

  // 4. Validar Código de Área (NXX)
  if (digits[0] === '0') {
    throw new Error('Area code cannot start with zero');
  }
  if (digits[0] === '1') {
    throw new Error('Area code cannot start with one');
  }

  // 5. Validar Código de Intercambio (NXX)
  if (digits[3] === '0') {
    throw new Error('Exchange code cannot start with zero');
  }
  if (digits[3] === '1') {
    throw new Error('Exchange code cannot start with one');
  }

  return digits;
}